File: phonegap-couchbase-ios/phonegap-couchbase-ios/Classes/CouchMover.m
//
//  CouchMover.m
//  CouchMover
//
//  Created by Kevin Malakoff on 6/11/11.
//  Copyright 2011 None.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not
// use this file except in compliance with the License. You may obtain a copy of
// the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// License for the specific language governing permissions and limitations under
// the License.

#import "CouchMover.h"
#import "NSData_Base64Extensions.h"             // for asBase64EncodedString for authenticaton

NSString * const CONTENT_TYPE_JSON              = @"application/json";
NSString * const CONTENT_TYPE_FORM              = @"multipart/form-data";
NSString * const COUCHAPP_LOADED_VERSION_DOC    = @"loaded_version";
NSString * const COUCHAPP_LOADED_VERSION_FIELD  = @"loaded_rev";
#define RESPONSE_DATA_OK(data)                  ([self responseDataHasFieldAndValue:data field:@"ok" value:@"true"])
#define RESPONSE_DATA_DB_EXISTS(data)           ([self responseDataHasField:data field:@"db_name"])
#define RESPONSE_DATA_DOC_EXISTS(data)          ([self responseDataHasField:data field:@"_id"])

@implementation CouchMover

@synthesize serverURL;
@synthesize databaseName;

@synthesize _credentialString;

///////////////////////
// Public Interface
///////////////////////
-(CouchMover*)init:(NSURL*)inServerURL serverCredential:(NSURLCredential*)inServerCredential databaseName:(NSString*)inDatabaseName; 
{
    self = [super init];
    if (self == nil)
        return nil;
    
    self.serverURL = inServerURL;
    self.couchappServerCredential = inServerCredential;
    self.databaseName = inDatabaseName;
    
    return self;
}

-(BOOL)ensureDatabaseExists
{
    NSString *databaseURL = [self urlToDatabase];
    NSData *responseData;
    
    /////////////////////////
    // check existence - GET
    /////////////////////////
    responseData = [self serverHTTPRequest:databaseURL httpMethod:@"GET" response:nil];
    
    // failed to connect to server or parse response
    if(!responseData) 
    {
        NSLog(@"Failed to connect to the server at: %@", databaseURL);
        return NO;
    }
    
    // the DB exists
    if(RESPONSE_DATA_DB_EXISTS(responseData))
        return YES;
    
    /////////////////////////
    // doesn't exist so create - PUT
    /////////////////////////
    responseData = [self serverHTTPRequest:databaseURL httpMethod:@"PUT" response:nil];
    return RESPONSE_DATA_OK(responseData);
}

-(BOOL)documentHasChanged:(NSString*)documentName version:(NSString*)version
{
    // call this first because loadDocument calls this
    [self ensureDatabaseExists];

    // no version so always load
    if(!version)
        return TRUE;
    
    // no change so do not upload
    NSString *currentVersion = [self getCurrentDocumentVersion:documentName];
    return(!currentVersion || ([currentVersion compare:version] != NSOrderedSame));
}

-(BOOL)loadDocument:(NSString*)documentName version:(NSString*)version getAppAsJSONDataBlock:(NSData* (^)())getAppAsJSONDataBlock
{
    // no change in the document
    if(![self documentHasChanged:documentName version:version])
        return YES;
    
    NSAutoreleasePool *pool = [[NSAutoreleasePool alloc] init];

    NSString *documentURL = [self urlToDocument:documentName];
    NSData *responseData;
    
    /////////////////////////
    // check existence of app - HEAD and if it does, DELETE it before update
    // NOTE: this is to avoid having to update the provided couchapp with the _rev parameter from the existing app 
    /////////////////////////

    // get the current revision if app exists
    NSHTTPURLResponse *response = nil;
    [self serverHTTPRequest:documentURL httpMethod:@"HEAD" response:&response];
    NSString *_revCurrent = response ? [[response allHeaderFields] valueForKey:@"Etag"] : nil; 

    // delete the current app
    if(_revCurrent)
        [self serverHTTPRequest_DeleteDocument:documentURL _revCurrent:_revCurrent];
    
    // create
    responseData = [self serverHTTPRequest:documentURL httpMethod:@"PUT" data:getAppAsJSONDataBlock() contentType:CONTENT_TYPE_FORM response:nil];

    BOOL success = RESPONSE_DATA_OK(responseData);
    if(!success)
        NSLog(@"Failed to create the couch app at URL: %@", documentURL);

    // update the version -> if there is no revision, skip and it will be loaded again next time
    else
        [self setCurrentDocumentVersion:documentName version:version];
    [pool release];
    
    return success;
}

-(BOOL)loadDocumentFromBundle:(NSBundle*)bundle documentName:(NSString*)documentName documentBundlePath:(NSString*)documentBundlePath versionBundlePath:(NSString*)versionBundlePath
{
    NSString *versionAbsolutePath = [[bundle resourcePath] stringByAppendingPathComponent:versionBundlePath];
    NSString *version = [NSString stringWithContentsOfFile:versionAbsolutePath usedEncoding:nil error:nil];
    
    if(!version)
    {
        NSLog(@"Failed to find the bundle resource: %@", versionAbsolutePath);
        return NO;
    }
    
    // load the coachapp if needed
    return [self loadDocument:documentName version:version getAppAsJSONDataBlock:^(){
        NSString *documentAbsolutePath = [[bundle resourcePath] stringByAppendingPathComponent:documentBundlePath];
        return (NSData*) [NSData dataWithContentsOfFile:documentAbsolutePath];
    }];
}

-(void)gotoAppPage:(NSString*)appDocumentName webView:(UIWebView*)webView page:(NSString*)page
{
    NSString *couchappApgeURL = [[NSString alloc] initWithFormat:@"window.location.href = \"%@/%@\"", [self urlToDocument:appDocumentName], page];
    [webView stringByEvaluatingJavaScriptFromString:couchappApgeURL];
    [couchappApgeURL release];
}

-(NSURLCredential*)couchappServerCredential
{
    return couchappServerCredential;
}

-(void)setCouchappServerCredential:(NSURLCredential*)inCredential
{
    [inCredential retain];
    
    if(couchappServerCredential)
        [couchappServerCredential release];
    
    couchappServerCredential = inCredential;

    if(couchappServerCredential)
    {
        NSString *tempString = [[NSString alloc] initWithFormat:@"%@:%@", couchappServerCredential.user, couchappServerCredential.password];
        NSData *credentialData = [tempString dataUsingEncoding:NSUTF8StringEncoding];
        [tempString release];
        tempString = [credentialData asBase64EncodedString:0];
        _credentialString = [[NSString alloc] initWithFormat:@"Basic %@", tempString];
    }
    else
    {
        [_credentialString release];
        _credentialString = nil;
    }
}


///////////////////////
// Internal Flow
///////////////////////
-(NSString*)getCurrentDocumentVersion:(NSString*)documentName
{
    NSString *documentURL = [self urlToLoadedDocumentVerionDocument:documentName];
    NSData *responseData;
    
    /////////////////////////
    // check existence - GET
    /////////////////////////
    responseData = [self serverHTTPRequest:documentURL httpMethod:@"GET" response:nil];
    
    // failed to connect to server or parse response
    if(!responseData) 
        return nil;

    // get the key field key
    return RESPONSE_DATA_DOC_EXISTS(responseData)? [self responseDataExtractSimpleValue:responseData field:COUCHAPP_LOADED_VERSION_FIELD] : nil;
}

-(void)setCurrentDocumentVersion:(NSString*)documentName version:(NSString*)version
{
    NSString *documentURL = [self urlToLoadedDocumentVerionDocument:documentName];
    NSData *responseData;
    
    /////////////////////////
    // check existence - GET
    /////////////////////////
    responseData = [self serverHTTPRequest:documentURL httpMethod:@"GET" response:nil];
    
    // exists, delete
    if(RESPONSE_DATA_DOC_EXISTS(responseData))
    {
        NSString *_revCurrent = [self responseDataExtractSimpleValue:responseData field:@"_rev"];
        [self serverHTTPRequest_DeleteDocument:documentURL _revCurrent:_revCurrent];
    }
    
    NSString *versionDocument = [[NSString alloc] initWithFormat:@"{\"%@\":\"%@\"}", COUCHAPP_LOADED_VERSION_FIELD, version];
    NSData *versionData = [versionDocument dataUsingEncoding:NSUTF8StringEncoding];
    
    // create
    responseData = [self serverHTTPRequest:documentURL httpMethod:@"PUT" data:versionData contentType:CONTENT_TYPE_FORM response:nil];
    if(!RESPONSE_DATA_OK(responseData))
        NSLog(@"Failed to create the couch app version document at URL: %@", documentURL);
    
    [versionDocument release];
}

///////////////////////
// URL Helpers
///////////////////////
-(NSString*)urlToDatabase
{
    return [NSString stringWithFormat:@"%@%@", [self.serverURL absoluteString], self.databaseName];
}

-(NSString*)urlToDocument:(NSString*)documentName
{
    return [NSString stringWithFormat:@"%@%@/%@", [self.serverURL absoluteString], self.databaseName, documentName];
}

-(NSString*)urlToLoadedDocumentVerionDocument:(NSString*)documentName
{
    if([documentName hasPrefix:@"_design"])
        return [NSString stringWithFormat:@"%@%@/%@_%@", [self.serverURL absoluteString], self.databaseName, documentName, COUCHAPP_LOADED_VERSION_DOC];

    // make sure _design starts with _design so it is skipped in view generation
    else
        return [NSString stringWithFormat:@"%@%@/_design/%@_%@", [self.serverURL absoluteString], self.databaseName, documentName, COUCHAPP_LOADED_VERSION_DOC];
}

///////////////////////
// HTTP Helpers
///////////////////////
-(NSData*)serverHTTPRequest:(NSString*)urlString httpMethod:(NSString*)httpMethod response:(NSHTTPURLResponse**)response;
{
    return [self serverHTTPRequest:urlString httpMethod:httpMethod data:nil contentType:nil response:response];
}

-(NSData*)serverHTTPRequest:(NSString*)urlString httpMethod:(NSString*)httpMethod data:(NSData*)data contentType:(NSString*)contentType response:(NSHTTPURLResponse**)response
{
    NSMutableURLRequest *    request = [[NSMutableURLRequest alloc] init];
    [request setHTTPMethod:httpMethod];
    [request setValue:CONTENT_TYPE_JSON forHTTPHeaderField:@"Accept"];
    [self requestAddURLString:request urlString:urlString];
    [self requestAddCredential:request];
    
    if(data && contentType)
    {
        // TODO: add a proper referer header
        [request setValue:[[request URL] absoluteString] forHTTPHeaderField:@"Referer"];
        [request setValue:contentType forHTTPHeaderField:@"Content-Type"];
        [request setValue:[NSString stringWithFormat:@"%d", [data length]] forHTTPHeaderField:@"Content-Length"];
        [request setHTTPBody:data];
    }

    // check the couch server for the database
    NSError *error = nil;
    NSData *responseData = [NSURLConnection sendSynchronousRequest:request returningResponse:response error:&error];
    //    [request release];
    if (error)
    {
        NSLog(@"Error: %@ \nFor request: %@",[error localizedDescription], [[request URL] absoluteString]);
        return nil;
    }
    
    return responseData;
}
-(BOOL)serverHTTPRequest_DeleteDocument:(NSString*)urlString _revCurrent:(NSString*)_revCurrent
{
    // generate the URL including the rev parameter (removing the quotes)
    NSString *documentURLDelete = [NSString stringWithFormat:@"%@?rev=%@", urlString, [_revCurrent stringByReplacingOccurrencesOfString:@"\"" withString: @""]];
    NSData *responseData = [self serverHTTPRequest:documentURLDelete httpMethod:@"DELETE" response:nil];
    if(!RESPONSE_DATA_OK(responseData))
    {
        NSLog(@"Failed to delete the at URL: %@", documentURLDelete);
        return NO;
    }
    
    return YES;
}

-(void)requestAddURLString:(NSMutableURLRequest*)request urlString:(NSString*)urlString
{
    NSURL *url = [NSURL URLWithString:urlString];
    [request setURL:url];
}

-(void)requestAddCredential:(NSMutableURLRequest*)request
{
    if(_credentialString)
        [request setValue:_credentialString forHTTPHeaderField:@"Authorization"];
}

-(BOOL)responseDataHasFieldAndValue:(NSData*)responseData field:(NSString*)field value:(NSString*)value
{
    NSString *valueString = [self responseDataExtractSimpleValue:responseData field:field];
    return (valueString && ([valueString compare:value] == NSOrderedSame));
}

-(BOOL)responseDataHasField:(NSData*)responseData field:(NSString*)field
{
    NSString *responseDataAsString = [[NSString alloc] initWithData:responseData encoding:NSUTF8StringEncoding];
    NSUInteger responseDataAsStringLength = [responseDataAsString length];
    
    NSString *regexPattern = [[NSString alloc] initWithFormat:@"\"%@\"( )*:", field];
    NSRegularExpression *fieldRegex = [[NSRegularExpression alloc] initWithPattern:regexPattern options:0 error:nil];
    [regexPattern release];
    NSTextCheckingResult *fieldMatch = [fieldRegex firstMatchInString:responseDataAsString options:0 range:NSMakeRange(0, responseDataAsStringLength)];
    [fieldRegex release];

    [responseDataAsString release];
    return (fieldMatch != nil);
}

// this is not robust -> the "field" should only occur once and the value should be a simple value (',' or '}')
-(NSString*)responseDataExtractSimpleValue:(NSData*)responseData field:(NSString*)field
{
    NSString *valueString = nil;
    NSString *responseDataAsString = [[NSString alloc] initWithData:responseData encoding:NSUTF8StringEncoding];
    NSUInteger responseDataAsStringLength = [responseDataAsString length];
    
    NSString *regexPattern = [[NSString alloc] initWithFormat:@"\"%@\"( )*:", field];
    NSRegularExpression *fieldRegex = [[NSRegularExpression alloc] initWithPattern:regexPattern options:0 error:nil];
    [regexPattern release];
    NSTextCheckingResult *fieldMatch = [fieldRegex firstMatchInString:responseDataAsString options:0 range:NSMakeRange(0, responseDataAsStringLength)];
    [fieldRegex release];
    
    // found the field
    if(fieldMatch)
    {
        NSRange fieldMatchRange = [fieldMatch range];
        NSUInteger fieldColonOffset = fieldMatchRange.location + fieldMatchRange.length;
        
        NSCharacterSet *fieldEndDelimiters = [NSCharacterSet characterSetWithCharactersInString:@"\\},"]; 
        NSRange valueMatchRange = [responseDataAsString rangeOfCharacterFromSet:fieldEndDelimiters options:0 range:NSMakeRange(fieldColonOffset, responseDataAsStringLength - fieldColonOffset)];

        // found the value
        if(valueMatchRange.length>0)
        {
            valueMatchRange.length = valueMatchRange.location - fieldColonOffset;       // drop the , or }
            valueMatchRange.location = fieldColonOffset;
            valueMatchRange = [self valueClean:responseDataAsString range:valueMatchRange];
            
            valueString = [responseDataAsString substringWithRange:valueMatchRange];
        }
    }

    [responseDataAsString release];
    return valueString;
}

///////////////////////
// Misc Helpers
///////////////////////
-(NSRange)valueClean:(NSString*)valueString range:(NSRange)range
{
    NSRange cleanRange = range;
    
    // remove leading space
    while([valueString characterAtIndex:cleanRange.location]==' ')
    {
        cleanRange.location++;
        cleanRange.length--;
    }
    
    // remove leading quote
    if([valueString characterAtIndex:cleanRange.location]=='"')
    {
        cleanRange.location++;
        cleanRange.length--;
    }

    // remove trailling space
    while([valueString characterAtIndex:cleanRange.location+cleanRange.length-1]==' ')
        cleanRange.length--;
    
    // remove trailing quote
    if([valueString characterAtIndex:cleanRange.location+cleanRange.length-1]=='"')
        cleanRange.length--;
    
    return cleanRange;
}

- (void)dealloc
{
    [_credentialString release];

	[serverURL release];
	[couchappServerCredential release];
	[databaseName release];
	[super dealloc];
}

@end

File: phonegap-couchbase-android/src/com/couchbaseextensions/CouchMover.java

package com.couchbaseextensions;

import java.io.BufferedReader;
import java.io.ByteArrayInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.PasswordAuthentication;
import java.net.URL;
import java.util.List;
import java.util.Map;
import java.util.Scanner;

import org.json.JSONException;
import org.json.JSONObject;

import android.content.res.AssetManager;
import android.util.Base64;
import android.util.Log;

public class CouchMover
{
	public interface LoaderCallback { public InputStream loadAsInputStream(); }
    final class AssetManagerDocumentLoaderCallback implements LoaderCallback
    {
    	private AssetManager assetManager; private String documentPath;
    	public AssetManagerDocumentLoaderCallback(AssetManager inAssetManager, String inDocumentPath) { assetManager = inAssetManager; documentPath = inDocumentPath; }
		public InputStream loadAsInputStream() { return CouchMover.readAssetAsInputStream(assetManager, documentPath);}
	};

	private String serverURLString;
	private String databaseName;

	protected String _credentialString;
	
	protected static final String LOG_TAG = "CouchMover";
	private class HTTPURLResponse 
	{
		public String responseString; public JSONObject json; public int statusCode; public Map<String, List<String>> headers;
		public HTTPURLResponse(String responseString, JSONObject json, int statusCode, Map<String, List<String>> headers) { this.responseString = responseString; this.json = json; this.statusCode = statusCode; this.headers = headers;}
	};

	protected static final String ACCEPT_TYPE_JSON              		= "application/json";
	protected static final String CONTENT_TYPE_JSON              		= "application/json; charset=UTF-8";
	protected static final String CONTENT_TYPE_FORM              		= "multipart/form-data";
	protected static final String COUCHAPP_LOADED_VERSION_DOC    		= "loaded_version";
	protected static final String COUCHAPP_LOADED_VERSION_FIELD  		= "loaded_rev";
	private boolean RESPONSE_DATA_OK(HTTPURLResponse data)				{return responseDataHasFieldAndValue(data, "ok", "true");}
	private boolean RESPONSE_DATA_DB_EXISTS(HTTPURLResponse data)		{return responseDataHasField(data, "db_name");}
	private boolean RESPONSE_DATA_DOC_EXISTS(HTTPURLResponse data)		{return responseDataHasField(data, "_id");}

	public CouchMover(String inServerURLString, PasswordAuthentication inServerCredential, String inDatabaseName) 
	{
		serverURLString = (inServerURLString!=null) ? new String(inServerURLString) : null;
		databaseName = (inDatabaseName!=null) ? new String(inDatabaseName) : null;
		_credentialString = null;
		setServerCredential(inServerCredential);
    }
	
	public void setDatabaseName(String inDatabaseName)
	{
		databaseName = new String(inDatabaseName);
	}
	
	public void setServerCredential(PasswordAuthentication inServerCredential)
	{
	    if(inServerCredential != null)
	    {
	    	String tempString = String.format("%s:%s", inServerCredential.getUserName(), inServerCredential.getPassword());
	        _credentialString = "Basic "+ Base64.encodeToString(tempString.getBytes(),Base64.URL_SAFE|Base64.NO_WRAP);
	    }
	    else
	    {
	        _credentialString = null;
	    }
	}

    public boolean ensureDatabaseExists() 
	{
	    String databaseURL = urlToDatabase();
	    HTTPURLResponse responseData;
		
	    /////////////////////////
	    // check existence - GET
	    /////////////////////////
	    responseData = serverHTTPURLResponse(databaseURL, "GET");
	    
	    // failed to connect to server or parse response
	    if(responseData==null) 
	    {
			Log.v(LOG_TAG, String.format("Failed to connect to the server at: %s", databaseURL));
	        return false;
	    }
	    
	    // the DB exists
	    if(RESPONSE_DATA_DB_EXISTS(responseData))
	        return true;
	    
	    /////////////////////////
	    // doesn't exist so create - PUT
	    /////////////////////////
	    responseData = serverHTTPURLResponse(databaseURL, "PUT");
	    return RESPONSE_DATA_OK(responseData);
	}

	public boolean documentHasChanged(String documentName, String version)
	{
	    // call this first because loadDocument calls this
	    ensureDatabaseExists();

	    // no version so always load
	    if(version==null)
	        return true;
	    
	    // no change so do not upload
	    String currentVersion = getCurrentDocumentVersion(documentName);
	    return((currentVersion == null) || (currentVersion.compareTo(version) != 0));
	}

	public boolean loadDocument(String documentName, String version, LoaderCallback documentLoaderCallback)
	{
	    // no change in the document
	    if(!documentHasChanged(documentName,version))
	        return true;
	    
	    String documentURL = urlToDocument(documentName);
	    HTTPURLResponse responseData;
	    
	    /////////////////////////
	    // check existence of document - HEAD and if it does, DELETE it before update
	    // NOTE: this is to avoid having to update the provided document with the _rev parameter from the existing document
	    /////////////////////////

	    // get the current revision if document exists
	    responseData = serverHTTPURLResponse(documentURL, "HEAD", null, null, true);
	    List<String> Values = ((responseData!=null) && (responseData.headers!=null)) ? responseData.headers.get("Etag") : null; 
	    String _revCurrent = (Values!=null) && (!Values.isEmpty()) ? Values.get(0): null;
	    	
	    // delete the current document
	    if(_revCurrent != null)
	        serverHTTPURLResponse_DeleteDocument(documentURL, _revCurrent);
	    
	    // create
	    responseData = serverHTTPURLResponse(documentURL, "PUT", documentLoaderCallback.loadAsInputStream(), CONTENT_TYPE_FORM, false);

	    boolean success = RESPONSE_DATA_OK(responseData);
	    if(!success)
	        Log.v(LOG_TAG, String.format("Failed to create the document at URL: %s", documentURL));

	    // update the version -> if there is no revision, skip and it will be loaded again next time
	    else
	        setCurrentDocumentVersion(documentName, version);
	    
	    return success;
	}

	public boolean loadDocumentFromAssetManager(AssetManager assetManager, String documentName, String documentAssetPath, String versionAssetPath)
	{
	    String version = inputStreamToString(CouchMover.readAssetAsInputStream(assetManager, versionAssetPath));
	    
	    if(version==null)
	    {
	        Log.v(LOG_TAG, String.format("Failed to find the bundle resource: %s", versionAssetPath));
	        return false;
	    }
		
	    // load the document if needed
	    return loadDocument(documentName, version, new AssetManagerDocumentLoaderCallback(assetManager, documentAssetPath));
	}

	public static InputStream readAssetAsInputStream(AssetManager assetManager, String assetPath)
	{
		try 
		{
			return assetManager.open(assetPath);
		} 
		catch (IOException e) 
		{
			e.printStackTrace();
		}
		
		return null;
	}
	
	public static String inputStreamToString(InputStream inputStream)
	{
		if(inputStream==null)
			return null;
		
		return new Scanner(inputStream, "UTF-8").useDelimiter("\\A").next();
	}

	public static void copyInputToOutputStream(InputStream inputStream, OutputStream outputStream)
	{
		final int MAX_COPY_BUFFER_SIZE = 64 * 1024; 
		
		try 
		{
			int bufferSize = inputStream.available();
			if(bufferSize>MAX_COPY_BUFFER_SIZE)
				bufferSize = MAX_COPY_BUFFER_SIZE;
			byte[] buffer = new byte[bufferSize];
			int length;

			while(true) 
			{
				length = inputStream.read(buffer);
				if (length < 0)
					break;
				outputStream.write(buffer, 0, length);
			}
		
			// close it, this is implicit knowledge that the input buffers are owned by the caller of this function and hopefully there are no memory leaks for failure conditions
			inputStream.close();
		} 
		catch (IOException e) 
		{
		}
	}
	
    public void gotoAppPage(String appDocumentName, android.webkit.WebView webView, String page)
	{
	    String couchappAppPageURL = String.format("%s/%s", urlToDocument(appDocumentName), page);
	    webView.loadUrl(couchappAppPageURL);
	}

	///////////////////////
	// Internal Flow
	///////////////////////
	private String getCurrentDocumentVersion(String documentName)
	{
	    String documentURL = urlToLoadedDocumentVerionDocument(documentName);
	    HTTPURLResponse responseData;
	    
	    /////////////////////////
	    // check existence - GET
	    /////////////////////////
	    responseData = serverHTTPURLResponse(documentURL, "GET");
	    
	    // failed to connect to server or parse response
	    if(responseData==null) 
	        return null;

	    // get the key field key
	    return RESPONSE_DATA_DOC_EXISTS(responseData)? responseDataExtractStringValue(responseData, COUCHAPP_LOADED_VERSION_FIELD) : null;
	}

	private void setCurrentDocumentVersion(String documentName, String version)
	{
	    String documentURL = urlToLoadedDocumentVerionDocument(documentName);
	    HTTPURLResponse responseData;
	    
	    /////////////////////////
	    // check existence - GET
	    /////////////////////////
	    responseData = serverHTTPURLResponse(documentURL, "GET");
	    
	    // exists, delete
	    if(RESPONSE_DATA_DOC_EXISTS(responseData))
	    {
	        String _revCurrent = responseDataExtractStringValue(responseData, "_rev");
	        serverHTTPURLResponse_DeleteDocument(documentURL, _revCurrent);
	    }
	    
	    String versionDocument = String.format("{\"%s\":\"%s\"}", COUCHAPP_LOADED_VERSION_FIELD, version);
	    
	    // create
	    responseData = serverHTTPURLResponse(documentURL, "PUT", new ByteArrayInputStream(versionDocument.getBytes(), 0, versionDocument.length()), CONTENT_TYPE_FORM, false);
	    if(!RESPONSE_DATA_OK(responseData))
	        Log.v(LOG_TAG, String.format("Failed to create the document version document at URL: %s", documentURL));
	}

	///////////////////////
	// URL Helpers
	///////////////////////
	private String urlToDatabase()
	{
	    return String.format("%s%s", serverURLString, databaseName);
	}

	private String urlToDocument(String documentName)
	{
	    return String.format("%s%s/%s", serverURLString, databaseName, documentName);
	}

	private String urlToLoadedDocumentVerionDocument(String documentName)
	{
	    if(documentName.startsWith("_design"))
	        return String.format("%s%s/%s_%s", serverURLString, databaseName, documentName, COUCHAPP_LOADED_VERSION_DOC);

	    // make sure _design starts with _design so it is skipped in view generation
	    else
	        return String.format("%s%s/_design/%s_%s", serverURLString, databaseName, documentName, COUCHAPP_LOADED_VERSION_DOC);
	}

	///////////////////////
	// HTTP Helpers
	///////////////////////
	private HTTPURLResponse serverHTTPURLResponse(String urlString, String httpMethod)
	{
		return serverHTTPURLResponse(urlString, httpMethod, null, null, false);
	}

	private HTTPURLResponse serverHTTPURLResponse(String urlString, String httpMethod, InputStream dataStream, String contentType, boolean includeHeaders)
	{
		String responseString = null;
		int statusCode = 0;
		Map<String, List<String>> headers = null;

		try {

			HttpURLConnection connection = (HttpURLConnection) new URL(urlString).openConnection();

			connection.setDoOutput(true);
			connection.setUseCaches(false);

			connection.setRequestMethod(httpMethod);
			connection.setRequestProperty("Accept", ACCEPT_TYPE_JSON);
			
			if(_credentialString!=null)
				connection.setRequestProperty(_credentialString, "Authorization");

			if((dataStream!=null)&&(contentType!=null))
			{
				connection.setDoInput(true);
				connection.setRequestProperty("Referer", urlString);
				connection.setRequestProperty("Content-type", contentType);
				connection.setRequestProperty("Content-Length", Integer.toString(dataStream.available()));
				copyInputToOutputStream(dataStream, connection.getOutputStream());
			}
			else
				connection.setRequestProperty("Content-type", CONTENT_TYPE_JSON);

			// call and get response
			connection.connect();
			statusCode = connection.getResponseCode();

			try 
			{
				// HEAD response is header only
				if(httpMethod!="HEAD")
				{
				    BufferedReader rd = new BufferedReader(new InputStreamReader(connection.getInputStream()));
					StringBuilder sb = new StringBuilder();
				    String line = null;
			        
			        while ((line = rd.readLine()) != null)
			        {
			        	sb.append(line);
			        }
					rd.close();
					responseString = sb.toString();
	
					//Log.v(LOG_TAG, String.format("Response string: %s", responseString));
				}				
			} catch (FileNotFoundException e) {

			} catch (NullPointerException e) {

			}

			finally {
				if((httpMethod=="HEAD") || (includeHeaders))
					headers = connection.getHeaderFields();
				
				connection.disconnect();
			}

		} catch (IOException e) {
			e.printStackTrace();
		}

		try
		{
			JSONObject json = ((responseString==null) || (responseString.length() == 0))
				? new JSONObject()
				: new JSONObject(responseString);

			return new HTTPURLResponse(responseString, json, statusCode, headers);
		} 
		catch (JSONException e) 
		{
			e.printStackTrace();
		}
		
		return null;
	}
	
	private boolean serverHTTPURLResponse_DeleteDocument(String urlString, String _revCurrent)
	{
	    // generate the URL including the rev parameter (removing the quotes)
	    String documentURLDelete = String.format("%s?rev=%s", urlString, _revCurrent.replaceAll("\"", ""));
	    HTTPURLResponse responseData = serverHTTPURLResponse(documentURLDelete, "DELETE");
	    if(!RESPONSE_DATA_OK(responseData))
	    {
	        Log.v(LOG_TAG, String.format("Failed to delete the at URL: %s", documentURLDelete));
	        return false;
	    }
	    
	    return true;
	}

	boolean responseDataHasFieldAndValue(HTTPURLResponse data, String field, String value)
	{
		try 
		{
			return (data.json.getString(field).compareTo(value)==0);
		} 
		catch (JSONException e) 
		{
			// OK, it may not exist
		}
		
		return false;
	}

	boolean responseDataHasField(HTTPURLResponse data, String field)
	{
		try 
		{
			return (data.json.get(field)!=null);
		} 
		catch (JSONException e) 
		{
			// OK, it may not exist
		}
		
		return false;
	}

	String responseDataExtractStringValue(HTTPURLResponse data, String field)
	{
		try 
		{
			return data.json.getString(field);
		} 
		catch (JSONException e) 
		{
			// OK, it may not exist
		}
		
		return null;
	}
}
    
